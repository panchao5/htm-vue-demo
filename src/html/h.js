import * as vue from "vue";

function h(type, props, ...children) {
  const newProps = {};

  const directives = [];

  for (const key in props) {
    if (props.hasOwnProperty(key)) {
      const result = /^v-(?<directiveName>[a-zA-Z_][0-9a-zA-Z_]*)$/.exec(key);
      if (!result) {
        newProps[key] = props[key];
      } else {
        const { directiveName } = result.groups;
        const directive =
          directiveName === "show"
            ? vue.vShow
            : vue.resolveDirective(directiveName);
        if (directive) {
          directives.push([directive, props[key]]);
        }
      }
    }
  }

  let slots = props?.["v-slots"];

  if (children.length === 1) {
    if (typeof children[0] === "function") {
      slots = slots ?? {};

      if (!slots.hasOwnProperty("default")) {
        slots.default = children[0];
      }
    }
  } else if (children.length === 0) {
    return vue.withDirectives(vue.h(type, newProps), directives);
  }

  return vue.withDirectives(
    vue.h(type, newProps, slots ? slots : children),
    directives
  );
}
export default h;
