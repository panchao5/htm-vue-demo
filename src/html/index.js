import htm from 'htm';
import h from './h';

const html = htm.bind(h)

export default html;