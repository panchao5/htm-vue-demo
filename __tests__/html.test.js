import {
  defineComponent,
  Fragment,
  Directive,
} from "vue";
import { mount } from "@vue/test-utils";
import html from "../src/html";

jest.spyOn(global.console, "log")

describe("html", () => {
  test("returns vnode", () => {
    const wrapper = mount(() => html`<div>Hello World</div>`);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("supports Fragment", () => {
    const wrapper = mount(
      () =>
        html`<${Fragment}
          ><div>Hello World</div>
          <div>Hello World</div><//
        >`
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("supports custom components", () => {
    const TestComponent = defineComponent({
      template: "<div>Hello World</div>",
    });
    const wrapper = mount(() => html`<${TestComponent} />`);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("supports v-show", () => {
    const wrapper = mount(() => html`<div v-show=${false}>Hello World</div>`);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("suports custom directives", () => {

    /**
     * @type Directive<any, string>
     */
    const directive = {
      mounted(_, { value }) {
        console.log(`Hello ${value}`)
      }
    }

    const spy = jest.spyOn(global.console, "log").mockImplementation(() => {})
  
    const wrapper = mount(() => html`<div v-greeting="World">Hello World</div>`, {
      global: {
        directives: { 'greeting': directive }
      }
    });
    expect(spy).toHaveBeenCalledWith("Hello World")
    expect(wrapper.html()).toMatchSnapshot();

    spy.mockRestore();
  });

})





