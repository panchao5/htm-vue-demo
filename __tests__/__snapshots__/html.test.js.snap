// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html returns vnode 1`] = `"<div>Hello World</div>"`;

exports[`html suports custom directives 1`] = `"<div>Hello World</div>"`;

exports[`html supports Fragment 1`] = `
"<div>Hello World</div>
<div>Hello World</div>"
`;

exports[`html supports custom components 1`] = `"<div>Hello World</div>"`;

exports[`html supports v-show 1`] = `"<div style=\\"display: none;\\">Hello World</div>"`;
